File: go/internal/coagulator1d/naiveparallel/coagulator.go
package naiveparallel

import (
	"sync"

	"github.com/maxkuzn/advection-and-coagulation/algorithm/coagulation"

	"github.com/maxkuzn/advection-and-coagulation/internal/field1d"
)

type coagulator struct {
	base      coagulation.Coagulator
	batchSize int
}

func New(base coagulation.Coagulator, batchSize int) *coagulator {
	return &coagulator{
		base:      base,
		batchSize: batchSize,
	}
}

func (c *coagulator) Start() error {
	return nil
}

func (c *coagulator) Stop() error {
	return nil
}

func (c *coagulator) Process(field, buff field1d.Field) (field1d.Field, field1d.Field) {
	var wg sync.WaitGroup

	for i := 0; i < field.Len(); i += c.batchSize {
		startIdx := i
		endIdx := i + c.batchSize
		if endIdx > field.Len() {
			endIdx = field.Len()
		}

		wg.Add(1)
		go func() {
			defer wg.Done()

			for j := startIdx; j < endIdx; j++ {
				c.base.Process(field.Cell(j), buff.Cell(j), field.Volumes())
			}
		}()
	}

	wg.Wait()

	return field, buff
}

File: cpp/library/coagulator1d/parallelpool/coagulator.cpp
#include "coagulator.h"

#include <memory>
#include <vector>
#include <thread>
#include <iostream>

namespace coagulation {

std::pair<Field1D*, Field1D*> ParallelPoolCoagulator1D::Process(Field1D* field, Field1D* buff) {
    std::vector<std::thread> threads;
    threads.reserve(field->Size());

    for (size_t i = 0; i < field->Size(); i += batch_size_) {
        size_t start = i;
        size_t end = std::min(i + batch_size_, field->Size());

        std::unique_lock lock(mtx_);

        queue_.push(std::make_tuple(field, buff, start, end));

        cv_.notify_one();
    }

    for (auto&& t: threads) {
        t.join();
    }

    return std::make_pair(field, buff);
}

void ParallelPoolCoagulator1D::DoWork() {
    for (;;) {
        std::unique_lock lock(mtx_);
        cv_.wait(lock, [&] {
            return stop_ || !queue_.empty();
        });

        if (stop_) {
            return;
        }

        auto [field, buff, start, end] = queue_.front();
        queue_.pop();

        lock.unlock();

        Field1D& f = *field;
        Field1D& b = *buff;


        for (size_t i = start; i < end; i++) {
            base_coagulator_->Process(&f[i], &b[i], volumes_);
        }
    }
}

};  // namespace coagulation
